pkg/dockerauth/multi.go

```diff
@@ -0,0 +1,32 @@
+package dockerauth
+
+import (
+	"fmt"
+
+	"github.com/fsouza/go-dockerclient"
+)
+
+type multiAuthProvider struct {
+	providers []AuthProvider
+}
+
+func NewMultiAuthProvider(providers ...AuthProvider) *multiAuthProvider {
+	return &multiAuthProvider{
+		providers: providers,
+	}
+}
+
+func (p *multiAuthProvider) GetAuthConfiguration(registry string) (docker.AuthConfiguration, error) {
+	for _, provider := range p.providers {
+		authConf, err := provider.GetAuthConfiguration(registry)
+		if err == nil {
```

Comment:
I think there's actually 2 concepts of a "Multi" provider here:

1. One where we attempt to find the first AuthConfiguration that matches the registry.
2. One where we degrade gracefully in the case of errors.

I feel like no 1. is what we're mostly trying to achieve here so I actually think changing the return value to `(*docker.AuthConfiguration, error)` would be the best option, and then return `nil, nil` from the ECR provider if the regex doesn't match, and `nil, nil` from the dockerConfig provider if the registry isn't found in the map.

For example, if ECR goes down and I'm deploying an ECR image, seeing `No auth provider can provide for dkr....` would be confusing to me as a user. I don't necessarily expect it to try to fall through to the dockerConfig provider because it's guaranteed to not work.

Hopefully that makes sense.